### PR TITLE
Implement Error::source on IntoStringError + Remove superfluous cause impls

### PR DESCRIPTION
**Pull request by faern**
IntoStringError only implemented `Error::cause`, which is deprecated. This implemements `Error::source` instead.
`Error::cause` will still work as before, thanks to the default implementation.

I think this was the only/last `Error` impl in the standard library to have a cause, but not a source.

### CODE REVIEW

**Comment by faern** [comment]
> I think this was the only/last Error impl in the standard library to have a cause, but not a source.

This statement was wrong. `TryLockError` also implements `cause` but no `source`. In this case it's harder to fix, because it has a generic parameter `T` which does not have the `'static` bound. So just implementing `source` does not work. And adding the bound breaks a lot of places where it's used.

**Comment by faern** [comment]
Pushed two more commits. Removing `cause` implementations that are not needed any more. Because they will be covered by `cause` forwarding to `source` by default.

File: src/libstd/io/error.rs [hunk 1]
```diff
@@ -548,15 +548,6 @@ impl error::Error for Error {
         }
     }
 
-    #[allow(deprecated)]
-    fn cause(&self) -> Option<&dyn error::Error> {
-        match self.repr {
-            Repr::Os(..) => None,
-            Repr::Simple(..) => None,
-            Repr::Custom(ref c) => c.error.cause(),
-        }
```
Comment (bluss) [inline src/libstd/io/error.rs]: This risks being a noticeable change since we are calling out to a user-supplied .cause() vs .source() here.

Comment (faern) [reply src/libstd/io/error.rs]: Good catch! I almost expected there was some aspect of this I missed :sweat_smile: 
I removed the last commit.

**Comment by bluss** [comment]
@bors r+ rollup

Thanks!

**Pull request merged**